#### Week 12 Exercise 12.2 Author: Rex Gayas Course & Section: DSC320-T301 Math for Data Science (2243-1) Date: 02 March 2024

#### Calculating Probabilities using Bayes’ Theorem 

In [4]:
# Solution to 1a.
# Given probabilities
P_A = 0.5
P_Def_A = 0.03
P_B = 0.3
P_Def_B = 0.02
P_C = 0.2
P_Def_C = 0.04

# Calculate P(Defective)
P_Defective = (P_A * P_Def_A) + (P_B * P_Def_B) + (P_C * P_Def_C)

# Calculate P(A|Defective) using Bayes' Theorem
P_A_Given_Defective = (P_Def_A * P_A) / P_Defective

# Convert the probability to a percentage
P_A_Given_Defective_Percentage = P_A_Given_Defective * 100

P_A_Given_Defective_Percentage

51.724137931034484

In [5]:
# Solution to 1b. 
# Given probabilities for non-defective
P_NotDef_A = 0.97
P_NotDef_B = 0.98
P_NotDef_C = 0.96

# Calculate P(Not Defective)
P_NotDefective = (P_A * P_NotDef_A) + (P_B * P_NotDef_B) + (P_C * P_NotDef_C)

# Calculate P(C|Not Defective) using Bayes' Theorem
P_C_Given_NotDefective = (P_NotDef_C * P_C) / P_NotDefective

# Convert the probability to a percentage
P_C_Given_NotDefective_Percentage = P_C_Given_NotDefective * 100

P_C_Given_NotDefective_Percentage


19.77342945417096

#### Entropy Function for a Probability Distribution 

In [7]:
# Solution to 2.
import math

def calculate_entropy(probabilities):
    """
    Calculate the entropy of a probability distribution.

    Parameters:
    probabilities (list): A list of probabilities for a discrete probability distribution.

    Returns:
    float: The entropy of the distribution.
    """
    # Calculate the entropy using the Shannon entropy formula
    entropy = -sum(p * math.log2(p) for p in probabilities if p > 0)
    return entropy


#### Calculating Entropies of Probability Distributions

In [8]:
# Solution to 3a. 
# Extracting the given probabilities for random variables X and Y 
probabilities_X = [0.2, 0.2, 0.2, 0.2, 0.2] # Given P(X = x)
probabilities_Y = [0.1, 0.4, 0.1, 0.3, 0.1] # Given P(Y = y)

# Using the previously defined function to calculate the entropy of X and Y
entropy_X = calculate_entropy(probabilities_X)
entropy_Y = calculate_entropy(probabilities_Y)

entropy_X, entropy_Y


(2.321928094887362, 2.046439344671015)

Solution to 3b.
The entropy of random variable X is approximately 2.322, and the entropy of random variable Y is approximately 2.046. 
The entropy of X is larger than the entropy of Y. Intuitively, this is because X has a uniform distribution, meaning each outcome is equally likely, which represents a state of maximum uncertainty and therefore maximum entropy. On the other hand, Y has a non-uniform distribution, with some outcomes being more probable than others, indicating less uncertainty and therefore lower entropy.